README.rst
claviger
********

``claviger`` manages the SSH ``authorized_keys`` files for you.

Quick introduction
==================

Tell ``claviger`` which keys you want to put on which server
by creating a ``~/.claviger``-file.  An example

.. code:: yaml

    keys:
        laptop: ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAINYZEwjtu8w9Hsvx85TlYE95MLV9Whc3N1ajrH7+gu7A
        desktop: ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAICUef9frJIX7tjvZkYYMtr4IdD/GcKz6/X5qvLxM1Z8O desktop
        work: ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAICrycv44eyFwWJ7QQsGOnjEiAsFSdxIoAEzBPSO/AQB5 work
    servers:
        myprivateserver.com:
            user: myusername
            present:
                - laptop
                - desktop
        root@myotherserver.com:
            like: myprivateserver.com
            present:
                - work
            keepOtherKeys: false
        workuser@workserver.com:
            present:
                - work
                - desktop
            absent:
                - laptop

Then run ``claviger``.  By default ``claviger`` only tells which changes
it wants to make, but does not make them.  If the changes seem fine,
run ``claviger -f``, which allows ``claviger`` to make changes.

Installation
============

To install ``claviger``, simply run::

   pip install claviger


Claviger config file
====================

A ``.claviger`` is written in YAML_.  It consists of two maps: the ``keys``
map and the ``servers`` map.

.. code:: yaml

    servers:
        server_key:
            # ... (server stanza)
        server_key2:
            # ... (server stanza)
    keys:
        key_name: # ... (ssh public key)
        key_name2: # ... (ssh public key)

As seen in the example at the top, the ``keys`` map has as values SSH
public keys as they would appear in an ``authorized_keys`` file.

The ``servers`` map consists of key-stanza pairs.  Each stanza is a map
of option name/value pairs, like:

.. code:: yaml

    server_key:
        option_name: option_value
        option_name2: option_value
        # ...

See below for the available options in the server-stanza's.

Server key
----------

The key of a server stanza is of the following form::

    [user@]hostname[:port]

Examples of keys are

- ``just-a-hostname.nl``
- ``user@some-server.com``
- ``some-server.nl:1234``
- ``user@and-port.com:22022``

You can also specify ``user``, ``hostname`` and ``port`` explicitly.
See below.

If a server key starts with a dollar sign (for instance ``$work``),
then it is considered ``abstract`` --- see below.

Server stanzas
--------------

A server stanza is a map which may have the following entries.

================== =============================================================
``name``           | The name of the server.
                   | *Default*: stanza key.
``hostname``       | The hostname of the server.
                   | *Default*: derived from stanza key.
``user``           | The user for which to manage the ``authorized_keys`` file
                   | *Default*: ``root`` if not derived from stanza key.
``present``        | A list of key names that must be in the
                     ``authorized_keys`` file.
                   | *Default*: the empty list ``[]``
``absent``         | A list of SSH-keys that should be removed from the
                     ``authorized_keys`` file.
                   | *Default*: the empty list ``[]``
``keepOtherKeys``  | ``true`` or ``false``.  If set to ``false``, ``claviger``
                     will remove all keys not explicitly allowed form the
                     ``authorized_keys`` file.
                   | *Default*: ``true``.
``allow``          | A list of SSH-keys that are also allowed to be in the
                     ``authorized_keys`` file if ``keepOtherKeys`` is set
                     to ``false``.  These keys will not be added, if
                     not present already.
                   | *Default*: the empty list ``[]``
``like``           | Name of another server stanza.  If set, the entries of
                     the other server stanza will be used as default values
                     for this server stanza.
                   | *Default*: ``$default``
``ssh_user``       | The user to use to get and put the
                     ``authorized_keys`` file.
                   | *Default*: the same as ``user``
``port``           | The port to use to connect to the server.
                   | *Default*: 22.
``abstract``       | ``true`` or ``false``. If set to ``true``, ``claviger``
                     will not check this server.  See below.
                   | *Default*: ``false``
================== =============================================================


Abstract servers and ``$default``
---------------------------------

``claviger`` will not check an *abstract* server.  This is useful to cleanly
configure multiple server.

.. code:: yaml

    servers:
        $mine:
            keepOthers: false
            present:
                - my_first_key
                - my_second_key
        $work:
            present:
                - my_work_key
            absent:
                - my_first_key

        my-first-server.tld:
            like: $mine
        my-second-server.tld:
            like: $mine
        alpha.at-work.tld:
            like: $work
        beta.at_work.tld:
            like: $work

By default, server inherits from the hidden ``$default`` abstract server.

.. code:: yaml

    servers:
        $default:
            user: myname
            present:
                - this_key_is_put_everywhere
        host1.tld: # will use myname as user
        host2.tld: # "
        root@host3.tld  # will use root as user
        host4.tld:
            absent:
                - this_key_is_put_everywhere # except here


.. _YAML: http://yaml.org


CHANGES.rst
claviger Changelog
******************

0.2.2 (unreleased)
==================

- Do not present configuration file errors as unhandled exceptions.


0.2.1 (2016-03-15)
==================

- Add missing dependency to jsonschema.  (Thanks-to: Joost Rijneveld)
- Properly decode output of `scp`.  (Thanks-to: Joost Rijneveld)


0.2 (2016-01-14)
================

- Add ``allow``-list to server stanza.

  If ``keepOtherKeys`` is set to false, claviger will remove any keys present
  except for those in the ``present`` and ``allow`` list.
- Add *abstract servers*.  Let every server inherit from the hidden
  abstract server ``$default``.
- Allow

  .. code:: yaml

    servers:
        server1:
        server2:

  which is prettier than

  .. code:: yaml

    servers:
        server1: {}
        server2: {}

- bugfix: actually set server name with ``name`` option.


0.1.2 (2016-01-08)
==================

- Show the correct number of keys actually removed.


0.1.1 (2015-12-21)
==================

- Some cosmetic changes.
- Show example configuration file, if claviger isn't configured yet.
- Python 3 compatibility.

0.1.0 (2015-12-14)
==================

- Initial release.


src/tests/test_authorized_keys.py
import unittest
import textwrap

import six

import claviger.authorized_keys

SSHD_MAN_PAGE_EXAMPLE = six.b(textwrap.dedent("""
    # Comments allowed at start of line 
    ssh-rsa AAAAB3NzaC1yc2EAAAA= user@example.net 
    from="*.sales.example.net,!pc.sales.example.net" ssh-rsa AAAAB3NzaC1yc2EAAAA= john@example.net 
    command="dump /home",no-pty,no-port-forwarding ssh-dss AAAAB3NzaC1kc3MAAA== example.net 
    permitopen="192.0.2.1:80",permitopen="192.0.2.2:25" ssh-dss AAAAB3NzaC1kc3MAAA== 
    tunnel="0",command="sh /etc/netstart tun0" ssh-rsa AAAAB3NzaC1yc2EAAAA= jane@example.net 
    restrict,command="uptime" ssh-rsa AAAAB3NzaC1yc2EAAAA= user@example.net 
    restrict,pty,command="nethack" ssh-rsa AAAAB3NzaC1yc2EAAAA= user@example.net
    """))

GRAWITY_EXAMPLE = six.b('ssh-foo="echo \\"Here\'s ssh-rsa for you\\"" '+
                      'future-algo AAAAC2Z1dHVyZS1hbGdv X y z.')

class TestAuthorizedKeys(unittest.TestCase):
    def test_man_page_examples(self):
        f = six.BytesIO(SSHD_MAN_PAGE_EXAMPLE)
        ak = claviger.authorized_keys.parse(f, False)
        self.assertTrue(ak.contains(b'AAAAB3NzaC1yc2EAAAA='))
        self.assertTrue(ak.contains(b'AAAAB3NzaC1kc3MAAA=='))
        self.assertEqual(len(ak.entries), 7)

        g = six.BytesIO()
        ak.store(g)
        self.assertEqual(f.getvalue(), g.getvalue())

    def test_grawity_hard_line(self):
        e = claviger.authorized_keys.Entry.parse(GRAWITY_EXAMPLE)
        self.assertEqual(e.keytype, b'future-algo')

    def test_getters(self):
        f = six.BytesIO(SSHD_MAN_PAGE_EXAMPLE)
        ak = claviger.authorized_keys.parse(f, False)
        self.assertEqual(len(ak.entries), 7)
        self.assertEqual(ak.get(b'doesnotexist'), None)

    def test_remove(self):
        f = six.BytesIO(SSHD_MAN_PAGE_EXAMPLE)
        ak = claviger.authorized_keys.parse(f, False)
        self.assertTrue(ak.contains(b'AAAAB3NzaC1yc2EAAAA='))
        ak.remove(b'AAAAB3NzaC1yc2EAAAA=')
        self.assertFalse(ak.contains(b'AAAAB3NzaC1yc2EAAAA='))
        self.assertTrue(ak.contains(b'AAAAB3NzaC1kc3MAAA=='))
        ak.removeAllKeys()
        self.assertEqual(len(ak.entries), 0)

if __name__ == '__main__':
    unittest.main()


src/scp.py
import os.path
import logging
import tempfile
import subprocess

l = logging.getLogger(__name__)

class SCP(object):
    def connect(self, hostname, port,  ssh_user):
        return SCPSession(hostname, port, ssh_user)


def interpret_scp_error(exitcode, stderr, stdout):
    """ Interpret the output of `scp' and create a suitable exception """
    if 'Host key verification failed' in stderr:
        return HostKeyVerificationFailed()
    msg = 'scp failed: exitcode {0}'.format(exitcode)
    if stderr.strip():
        msg += '; stderr {0}'.format(repr(stderr))
    if stdout.strip():
        msg += '; stdout {0}'.format(repr(stdout))
    return SCPError(msg)

class SCPError(Exception):
    pass
class HostKeyVerificationFailed(SCPError):
    pass

class SCPSession(object):
    def __init__(self, hostname, port, ssh_user):
        self.hostname = hostname
        self.port = port if port else 22
        self.ssh_user = ssh_user
    def _path_for(self, user):
        # TODO read passwd or use SSH session to find home
        return os.path.join('~' + user, '.ssh', 'authorized_keys')
    def _scp(self, src, trg):
        cmd = ['scp', '-B', '-P', str(self.port), src, trg]
        l.debug('executing %s', cmd)
        p = subprocess.Popen(cmd, stdout=subprocess.PIPE,
                                  stderr=subprocess.PIPE)
        stdout_txt, stderr_txt = [x.decode('utf-8') for x in p.communicate()]
        if p.returncode != 0:
            raise interpret_scp_error(p.returncode, stderr_txt, stdout_txt)

    def get(self, user):
        with tempfile.NamedTemporaryFile() as tempf:
            # FIXME escaping
            # TODO check for error
            self._scp('{0}@{1}:{2}'.format(self.ssh_user, self.hostname,
                            self._path_for(user)), tempf.name)
            return tempf.read()
        # TODO create .ssh if it does not exist
        # TODO check permissions
        # TODO backup?
    def put(self, user, authorized_keys):
        with tempfile.NamedTemporaryFile() as tempf:
            tempf.write(authorized_keys)
            tempf.flush()
            self._scp(tempf.name, '{0}@{1}:{2}'.format(self.ssh_user,
                            self.hostname, self._path_for(user)))


src/main.py
import demandimport
demandimport.enable()

import multiprocessing.dummy
import itertools
import traceback
import argparse
import textwrap
import os.path
import logging
import sys
import os

import claviger.authorized_keys
import claviger.config
import claviger.worker
import claviger.scp

import six
import yaml

l = logging.getLogger(__name__)

class Claviger(object):
    """ main object for claviger """
    def main(self, args):
        try:
            self.parse_commandline_args()

            extra_logging_config = {}
            if self.args.verbosity >= 2:
                level = logging.DEBUG
                extra_logging_config['format'] = ('%(relativeCreated)d '+
                        '%(levelname)s %(name)s %(message)s')
            elif self.args.verbosity == 1:
                level = logging.INFO
            else:
                level = logging.WARNING
            logging.basicConfig(level=level, **extra_logging_config)

            if not os.path.exists(self.args.configfile):
                return self.show_configuration_instructions()
            self.cfg = claviger.config.load(self.args.configfile)
            self.check_servers()
        except claviger.config.ConfigurationError as e:
            s = 'There was a problem with your configfile {0}:\n {1}\n'
            sys.stderr.write(s.format(self.args.configfile, e))
            return 3
        except Exception:
            return self.handle_uncaught_exception()

    def check_servers(self):
        if self.args.parallel_connections == 1:
            # If we want one worker, the current thread will do just fine.
            the_map = itertools.imap
        else:
            # As check_server is iobound, threads are better than processes.
            pool = multiprocessing.dummy.Pool(
                            processes=self.args.parallel_connections)
            the_map = pool.imap_unordered

        global_changes = False
        errors_occured = False
        for ret in the_map(claviger.worker.check_server,
                (claviger.worker.Job(server=self.cfg['servers'][server_name],
                                    keys=self.cfg['keys'],
                                    dry_run=self.args.dry_run,
                                    no_diff=self.args.no_diff)
                         for server_name in self.cfg['servers']
                         if not self.cfg['servers'][server_name]['abstract'])):
            if not ret.ok:
                errors_occured = True
                if isinstance(ret.result,
                        claviger.scp.HostKeyVerificationFailed):
                    print("{0:<40} host key verification failed".format(
                                ret.server_name))
                else:
                    print("")
                    print("{0:<40} error".format(ret.server_name))
                    print("   {0}".format(ret.result))
                    print("")
                continue
            res = ret.result
            changes = any([res.n_keys_added, res.n_keys_removed])
            global_changes |= changes
            l.debug('        %s: done', ret.server_name)
            if not changes and not self.args.verbosity:
                continue
            print ("{0:<40} +{1:<2} -{2:<2} ?{3:<2}".format(ret.server_name,
                            res.n_keys_added, res.n_keys_removed,
                            res.n_keys_ignored))
        if not global_changes and not errors_occured:
            print('Everything is in order.')
        elif self.args.dry_run and not errors_occured:
            print('')
            print("This is a dry run: no changes have been made.")
            print("Rerun with `-f' to apply changes.")
        elif errors_occured:
            print('')
            print('Checking some servers failed.  See above.')
            if self.args.dry_run:
                print("This is a dry run: no changes have been made.")

    def find_ssh_pubkeys(self):
        """ Searches for SSH public keys in the user's homedir.

            Return a list of (unique_name, pubkey) pairs. """
        ret = []
        names = set()
        ssh_dir = os.path.expanduser("~/.ssh")
        if not os.path.isdir(ssh_dir):
            return ret
        for fn in os.listdir(ssh_dir):
            if not fn.startswith('id_') or not fn.endswith('.pub'):
                continue
            with open(os.path.join(ssh_dir, fn), 'rb') as f:
                raw_pubkey = f.readline()
                if not raw_pubkey:
                    continue
                raw_pubkey = raw_pubkey[:-1]
            try:
                pubkey = claviger.authorized_keys.Entry.parse(raw_pubkey)
            except claviger.authorized_keys.CouldNotParseLine:
                continue
            name = pubkey.comment if pubkey.comment else 'mykey'
            if name in names:
                i = 2
                while name + str(i) in names:
                    i += 1
                name = name + str(i)
            names.add(name)
            ret.append((name, raw_pubkey))
        return ret

    def show_configuration_instructions(self):
        keys = self.find_ssh_pubkeys()
        if not keys:
            keys = [('exampleKey', 'ssh-rsa AAAAHitsanexamplekey comment')]
        print(textwrap.dedent("""
                The configuration file {configfile} does not exist.

                An example config file is:

                    keys: """).format(configfile=self.args.configfile))
        for name, key in keys:
            print("        {0}: {1}".format(yaml_str(name), yaml_str(key)))
        print(textwrap.dedent("""
                    servers:
                        user@example.com:
                            present:""").replace('\n','\n    ')[1:])
        for name, keys in keys:
            print("               - {0}".format(yaml_str(name)))
        return 1

    def parse_commandline_args(self):
        parser = argparse.ArgumentParser(
                        description='Synchronize remote SSH authorized_keys')
        parser.add_argument('-c', '--configfile', metavar='PATH',
                                type=os.path.expanduser, default='~/.claviger',
                                help='Configuration file')
        parser.add_argument('--verbose', '-v', action='count',
                            dest='verbosity', default=0,
                    help='Add to increase make claviger chatty')
        parser.add_argument('--parallel-connections', '-p', metavar='N',
                                type=int, default=8,
                    help='Number of parallel connections')
        parser.add_argument('--apply-changes', '-f', action='store_false',
                            dest='dry_run',
                    help='Apply changes')
        parser.add_argument('--no-diff', '-s', action='store_true',
                    help='Do not show a diff during the dry run')
        self.args = parser.parse_args()

    def handle_uncaught_exception(self):
        sys.stderr.write('\n')
        sys.stderr.write('An unexpected exception occured:\n')
        sys.stderr.write('\n    ')
        sys.stderr.write(traceback.format_exc().replace('\n', '\n    '))
        sys.stderr.write('\n')
        sys.stderr.write('Please report this error:\n')
        sys.stderr.write('\n')
        sys.stderr.write('  https://github.com/bwesterb/claviger/issues\n')
        sys.stderr.write('\n')
        sys.stderr.flush()
        return 2

def yaml_str(s):
    """ Escapes the string for inclusion in YAML. """
    if not isinstance(s, six.string_types):
        s = s.decode('utf-8')
    ret = yaml.dump(s)
    if ret.endswith('\n'):
        ret = ret[:-1]
    if ret.endswith('\n...'):
        ret = ret[:-4]
    return ret

def entrypoint():
    """ entry-point of claviger """
    return Claviger().main(sys.argv[1:])

if __name__ == '__main__':
    entrypoint()


src/sftp.py
import os.path
import logging
import tempfile
import subprocess

l = logging.getLogger(__name__)

class SFTP(object):
    def connect(self, hostname, port,  ssh_user):
        return SFTPSession(hostname, port, ssh_user)


def interpret_scp_error(exitcode, stderr, stdout):
    """ Interpret the output of `scp' and create a suitable exception """
    if 'Host key verification failed' in stderr:
        return HostKeyVerificationFailed()
    msg = 'scp failed: exitcode {0}'.format(exitcode)
    if stderr.strip():
        msg += '; stderr {0}'.format(repr(stderr))
    if stdout.strip():
        msg += '; stdout {0}'.format(repr(stdout))
    return SFTPError(msg)

class SFTPError(Exception):
    pass
class HostKeyVerificationFailed(SFTPError):
    pass

class SFTPSession(object):
    def __init__(self, hostname, port, ssh_user):
        self.hostname = hostname
        self.port = port if port else 22
        self.ssh_user = ssh_user
    def _path_for(self, user):
        # TODO read passwd or use SSH session to find home
        return os.path.join('~' + user, '.ssh', 'authorized_keys')
    def _scp(self, src, trg):
        cmd = ['scp', '-B', '-P', str(self.port), src, trg]
        l.debug('executing %s', cmd)
        p = subprocess.Popen(cmd, stdout=subprocess.PIPE,
                                  stderr=subprocess.PIPE)
        stdout_txt, stderr_txt = p.communicate()
        if p.returncode != 0:
            raise interpret_scp_error(p.returncode, stderr_txt, stdout_txt)

    def get(self, user):
        with tempfile.NamedTemporaryFile() as tempf:
            # FIXME escaping
            # TODO check for error
            self._scp('{0}@{1}:{2}'.format(self.ssh_user, self.hostname,
                            self._path_for(user)), tempf.name)
            return tempf.read()
        # TODO create .ssh if it does not exist
        # TODO check permissions
        # TODO backup?
    def put(self, user, authorized_keys):
        with tempfile.NamedTemporaryFile() as tempf:
            tempf.write(authorized_keys)
            tempf.flush()
            self._scp(tempf.name, '{0}@{1}:{2}'.format(self.ssh_user,
                            self.hostname, self._path_for(user)))


src/worker.py
""" Contains the code to check the authorized_keys file on a server.
    
    We run several checks in parallel using the multiprocessing module.
    To avoid serialization (and other) issues, we separate the worker
    environment in this module  with the rest of the program. """

import sys
import difflib
import traceback
import collections

import six

import claviger.scp
import claviger.authorized_keys

# arguments send by the main process
Job = collections.namedtuple('Job',
                ('server', 'keys', 'dry_run', 'no_diff'))

# this is what we return
JobReturn = collections.namedtuple('JobReturn',
                ('server_name', 'ok', 'result'))
# if everything is ok, the result field is of the following type ...
JobResult = collections.namedtuple('JobResult',
                ('n_keys_added', 'n_keys_removed', 'n_keys_ignored'))
# ... otherwise it is an exception

def check_server(job):
    try:
        scp = claviger.scp.SCP()
        n_keys_removed = 0
        n_keys_added = 0
        n_keys_ignored = 0
        server = job.server
        conn = scp.connect(server['hostname'], server['port'],
                                    server['ssh_user'])

        # First pull the current authorized_keys
        original_raw_ak = conn.get(server['user'])
        ak = claviger.authorized_keys.parse(original_raw_ak)

        # Check which keys to add
        for key_name in server['present']:
            key = job.keys[key_name]
            # TODO update comment/options
            if ak.contains(key['key']):
                continue
            n_keys_added += 1
            ak.add(key['options'], key['keytype'], key['key'], key['comment'])

        # Check which keys to remove
        if server['keepOtherKeys']:
            for key_name in server['absent']:
                if not ak.contains(job.keys[key_name]['key']):
                    continue
                ak.remove(job.keys[key_name]['key'])
                n_keys_removed += 1
        allowed = {job.keys[key_name]['key']
                        for key_name in server['present'] + server['allow']}

        for entry in ak.entries:
            if entry.key in allowed:
                continue
            if not server['keepOtherKeys']:
                ak.remove(entry.key)
                n_keys_removed += 1
            else:
                n_keys_ignored += 1

        # Did things change?
        raw_ak = six.binary_type(ak)
        if n_keys_added or n_keys_removed:
            if job.dry_run:
                if not job.no_diff:
                    print(''.join(difflib.unified_diff(
                                original_raw_ak.decode('utf-8').splitlines(True),
                                raw_ak.decode('utf-8').splitlines(True),
                                server['name'])))
            else:
                conn.put(server['user'], raw_ak)

        return JobReturn(server_name=server['name'], ok=True,
                    result=JobResult(n_keys_added=n_keys_added,
                                     n_keys_removed=n_keys_removed,
                                     n_keys_ignored=n_keys_ignored))
    except claviger.scp.SCPError as e:
        return JobReturn(server_name=server['name'], ok=False, result=e)
    except Exception as e:
        # multiprocessing does not pass the stacktrace to the parent process.
        #   ( see http://stackoverflow.com/questions/6126007 )
        # Thus we force the stacktrace in the message.
        raise Exception(''.join(traceback.format_exception(*sys.exc_info())))


src/config.py
""" Reads claviger's configuration file. """
import yaml
import logging
import os.path
import textwrap
import itertools
import collections

import six
import tarjan
import jsonschema

import claviger.authorized_keys

class ConfigError(Exception):
    pass

ParsedServerKey = collections.namedtuple('ParsedServerKey',
                                    ('hostname', 'user', 'port', 'abstract'))

l = logging.getLogger(__name__)

# Schema for the configuration file.
_SCHEMA = None

def get_schema():
    global _SCHEMA
    if not _SCHEMA:
        l.debug('loading scheme ...')
        with open(os.path.join(os.path.dirname(os.path.realpath(__file__)),
                'config.schema.yml')) as f:
            _SCHEMA = yaml.load(f)
        l.debug('    ... done!')
    return _SCHEMA

class ConfigurationError(Exception):
    pass

def parse_server_key(key):
    """ Converts a server key like (root@host:1234 or just host)
                to a triplet (user, port, hostname, abstract) """
    port = None
    user = None
    abstract = False
    hostname = None
    if key.startswith('$'):
        abstract = True
    else:
        hostname = key
        if ':' in hostname:
            hostname, _port = hostname.rsplit(':', 1)
            port = int(_port)
        if '@' in hostname:
            user, hostname = hostname.split('@', 1)
    return ParsedServerKey(user=user, port=port, hostname=hostname,
                                abstract=abstract)

def load(path):
    """ Loads the configuration file.
    
        A lot of the work is done by YAML.  We validate the easy bits with
        a JSON schema. The rest by hand. """
    # TODO Cache schema and configuration file
    l.debug('loading configuration file ...')
    with open(path) as f:
        cfg = yaml.load(f)

    if not isinstance(cfg, dict):
        raise ConfigurationError('Configuration file is empty')

    l.debug('  - checking schema')
    # First small fixes which the schema can't handle
    cfg.setdefault('servers', {})
    cfg['servers'].setdefault('$default', {})
    for key in cfg['servers']:
        if cfg['servers'][key] is None:
            cfg['servers'][key] = dict()

    # Now check the schema
    jsonschema.validate(cfg, get_schema())
    # TODO format into pretty error message

    l.debug('  - processing keys')
    new_keys = {}
    cfg.setdefault('keys', {})
    for key_name, key in six.iteritems(cfg['keys']):
        # TODO handle error
        entry = claviger.authorized_keys.Entry.parse(key)
        new_key = {'key': entry.key,
                   'options': entry.options,
                   'comment': entry.comment,
                   'keytype': entry.keytype}
        new_keys[key_name] = new_key
    cfg['keys'] = new_keys

    l.debug('  - processing server stanza short-hands')
    new_servers = {}
    for server_key, server in six.iteritems(cfg['servers']):
        parsed_server_key = parse_server_key(server_key)
        server.setdefault('name', server_key)
        server_name = server['name']
        server.setdefault('port', parsed_server_key.port)
        server.setdefault('user', parsed_server_key.user)
        server.setdefault('hostname', parsed_server_key.hostname)
        server.setdefault('ssh_user', server['user'])
        server.setdefault('present', [])
        server.setdefault('absent', [])
        server.setdefault('allow', [])
        server.setdefault('keepOtherKeys')
        server.setdefault('like', '$default' if server_key != '$default'
                                        else None)
        server.setdefault('abstract', parsed_server_key.abstract)
        prabsent = frozenset(server['present']) & frozenset(server['absent'])
        if prabsent:
            raise ConfigurationError(
                "Keys {0} are required to be both present and absent on {1}"
                    .format(tuple(prabsent), server_name))
        ablow = frozenset(server['allow']) & frozenset(server['absent'])
        if ablow:
            raise ConfigurationError(
                "Keys {0} are listed allowed and absent on {1}"
                    .format(tuple(ablow), server_name))
        for key_name in itertools.chain(server['present'], server['absent'],
                                        server['allow']):
            if not key_name in cfg['keys']:
                "Key {0} (on {1}) does not exist".format(key_name, server_name)
        if server_name in new_servers:
            raise ConfigurationError(
                "Duplicate server name {0}".format(server_name))
        new_servers[server_name] = server
    cfg['servers'] = new_servers

    l.debug('  - resolving server stanza inheritance')
    # create dependancy graph and use Tarjan's algorithm to find a possible
    # order to evaluate the server stanzas.
    server_dg = {server_name: [server['like']] if server['like'] else []
                    for server_name, server in six.iteritems(cfg['servers'])}
    for server_cycle_names in tarjan.tarjan(server_dg):
        if len(server_cycle_names) != 1:
            raise ConfigurationError(
                    "There is a cyclic dependacy among the servers {0}".format(
                                server_cycle_names))
        target_server = cfg['servers'][server_cycle_names[0]]
        if not target_server['like']:
            continue
        if not target_server['like'] in cfg['servers']:
            pass
        source_server = cfg['servers'][target_server['like']]

        # First the simple attributes
        for attr in ('port', 'user', 'hostname', 'ssh_user',
                        'keepOtherKeys'):
            if attr in source_server:
                if target_server[attr] is None:
                    target_server[attr] = source_server[attr]

        # Now, the present/absent/allow lists
        for key in source_server['present']:
            if key in target_server['absent']:
                continue
            if key not in target_server['present']:
                target_server['present'].append(key)
        for key in source_server['absent']:
            if (key in target_server['present']
                    or key in target_server['allow']):
                continue
            if key not in target_server['absent']:
                target_server['absent'].append(key)
        for key in source_server['allow']:
            if key in target_server['absent']:
                continue
            if key not in target_server['allow']:
                target_server['allow'].append(key)

    l.debug('  - setting defaults on server stanzas')
    for server in six.itervalues(cfg['servers']):
        for attr, dflt in (('port', 22),
                           ('user', 'root'),
                           ('keepOtherKeys', True)):
            if server[attr] is None:
                server[attr] = dflt
        
    l.debug('         ... done')

    return cfg
